semgrep/semgrep/dependency_aware_rule.py

```diff
@@ -56,76 +57,74 @@ def parse_depends_on_yaml(
         # no dependencies to process, so skip
         return matches, []
 
-    dummy_match = RuleMatch(
-        rule_id=rule.id,
-        message=rule.message,
-        metadata=rule.metadata,
-        severity=rule.severity,
-        path=top_level_target,  # TODO: this should probably be the offending lockfile
-        fix=None,
-        fix_regex=None,
-        start=core.Position(0, 0, 0),
-        end=core.Position(0, 0, 0),
-        extra={},
-    )
-
-    try:
-        depends_on_entries = list(parse_depends_on_yaml(dependencies))
-        output = list(
-            dependencies_range_match_any(
-                depends_on_entries, find_and_parse_lockfiles(lockfile_target)
+    depends_on_entries = list(parse_depends_on_yaml(depends_on_keys))
+    final_matches: List[RuleMatch] = []
+
+    namespaces = {LANGUAGE_TO_NAMESPACE[l] for l in rule.languages}
+    dependencies: List[Tuple[Path, List[LockfileDependency]]] = [
+        dep for ns in namespaces for dep in dep_trie.all_deps[ns].items()
+    ]
+    for lockfile_path, deps in dependencies:
+        try:
+            output = list(
+                dependencies_range_match_any(depends_on_entries, lockfile_path, deps)
             )
-        )
-        output_for_json = [
-            {
-                "dependency_pattern": vars(dep_pat),
-                "found_dependency": vars(found_dep),
-                "lockfile": lockfile.name,
-            }
-            for dep_pat, found_dep, lockfile in output
-        ]
-        # Dependency checks found nothing, so rule should not match
-        if output == []:
-            final_matches = []
-
-        # Dependency checks found something, so rule should match
-        # and we should get a dummy match indicating the dependency match
-        else:
-            if matches == []:
-                dummy_match.extra["dependency_match_only"] = True
-                dummy_match.extra["dependency_matches"] = output_for_json
-                matches.append(dummy_match)
-            else:
-                for match in matches:
+            if not output:
+                continue
+
+            output_for_json = [
+                {
+                    "dependency_pattern": vars(dep_pat),
+                    "found_dependency": vars(found_dep),
+                    "lockfile": lockfile.name,
+                }
+                for dep_pat, found_dep, lockfile in output
+            ]
+
+            reachable = []
+            matches_remaining = []
+            for match in matches:
+                if lockfile_path in (dep_trie.find_dependencies(match.path) or {}):
                     match.extra["dependency_match_only"] = False
                     match.extra["dependency_matches"] = output_for_json
+                    reachable.append(match)
+                else:
+                    matches_remaining.append(match)
+            matches = matches_remaining
+            if not reachable:
+                dummy_match = RuleMatch(
```

Comment:
What is this dummy_match for?

Reply:
`dummy_match` was a bad name. If the rule had no pattern matches, but there was a dependency match, we want to create a match that points to the offending lockfile. Then the `dependency_match_only` field  added to `extra` specifies that this is what happened.